client/auth/edit-account.jsx

```diff
@@ -0,0 +1,194 @@
+)
+
+@form({
+  currentPassword: passwordValidator,
+  newPassword: newPasswordValidator,
+  confirmNewPassword: confirmNewPasswordValidator,
+})
+class AccountForm extends React.Component {
+  render() {
+    const { bindInput, onSubmit } = this.props
+    const textInputProps = {
+      autoCapitalize: 'off',
+      autoCorrect: 'off',
+      spellCheck: false,
+      tabIndex: 1,
+    }
+
+    return (
+      <form noValidate={true} onSubmit={onSubmit}>
+        <SubmitOnEnter />
+        <PasswordTextField
+          {...bindInput('currentPassword')}
+          label='Current password'
+          floatingLabel={true}
+          inputProps={textInputProps}
+        />
+        <PasswordTextField
+          {...bindInput('newPassword')}
+          label='New password'
+          floatingLabel={true}
+          inputProps={textInputProps}
+        />
+        <PasswordTextField
+          {...bindInput('confirmNewPassword')}
+          label='Confirm new password'
+          floatingLabel={true}
+          inputProps={textInputProps}
+        />
+      </form>
+    )
+  }
+}
+
+const LoadingArea = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  margin-bottom: 24px;
+`
+
+const ErrorText = styled(Subheading)`
+  margin: 0;
+  margin-bottom: 24px;
+  color: ${colorError};
+`
+
+const AccountContainer = styled.div`
+  display: flex;
+  align-items: flex-start;
+`
+
+const StyledAvatar = styled(Avatar)`
+  flex-shrink: 0;
+  width: 128px;
+  height: 128px;
+  margin-right: 16px;
+`
+
+const InfoContainer = styled.div`
+  flex-grow: 1;
+  display: flex;
+  flex-direction: column;
+`
+
+@connect(state => ({ auth: state.auth }))
+export default class EditAccount extends React.Component {
+  state = {
+    reqId: null,
+  }
+
+  _form = React.createRef()
+
+  componentDidUpdate(prevProps) {
+    const isPrevRequesting = prevProps.auth.authChangeInProgress
+    const isRequesting = this.props.auth.authChangeInProgress
+    const hasError = !!this.props.auth.lastFailure
+
+    if (isPrevRequesting && !isRequesting && !hasError) {
+      this.props.dispatch(closeDialog())
+    }
+  }
+
+  renderDialogContents() {
+    const { auth } = this.props
+
+    return (
+      <AccountContainer>
+        <StyledAvatar user={auth.user.name} />
+        <InfoContainer>
+          <AccountForm ref={this._form} model={{}} onSubmit={this.onSubmit} />
+        </InfoContainer>
+      </AccountContainer>
+    )
+  }
+
+  render() {
+    const { auth, onCancel } = this.props
+    const { reqId } = this.state
+    let loadingElem
+    let errorElem
+
+    if (auth.authChangeInProgress) {
+      loadingElem = (
+        <LoadingArea>
+          <LoadingIndicator />
+        </LoadingArea>
+      )
+    }
+
+    if (reqId && auth.lastFailure && auth.lastFailure.reqId === reqId) {
+      errorElem = (
+        <ErrorText>
+          There was an issue with updating your account. Please check your data and try again.
```

Comment:
Not a huge fan of this error message :)

It's safe here to tell someone if they've entered a wrong password (because they already know this user exists, they're logged into it). We probably want to show something specific for that, otherwise something like:

There was an issue updating your account. Please try again later.

Reply:
Oh, I wasn't sure about the security implications of that :)
I'll update it so it's clear when the entered current password is wrong. It would be nice if it were possible to even manually set the error of that text field in form (like what Discord does).

Reply:
I don't think a minlength validator on the current password makes much sense anyway, tbh, although adding custom errors when we have validations feels sorta weird. Might be better to try and build more support for that into the form stuff rather than just avoiding doing it on something that has validation functions.

Reply:
Actually, I was playing a bit with that too afterwards, but not really sure what's the best way to deal with it. Seems like the form would need a way to run some validations _after_ a submit was done? Which doesn't really make sense. Or it would need to expose a method which can set the internal error state for that field manually, similar to how `setInputValue` can be used to set its value.

Also, not having minLength validator does seem a bit weird though now that I've tried it, since the server will just return BadRequest in that case, which will display a generic "something went wrong message". I guess we can just not check that on the server either and instead just return "Incorrect password" instead.

Reply:
Perhaps some system where the submission function could return a list of validation failures, or notify the form of validation failures from the submission in some way?

As far as the minlength stuff, I mostly meant that it doesn't make sense *provided* that we're checking that the field has content at all. No content should still be checked (and/or block submission entirely), but I guess that's still a validator that you're going to have trouble with. Returning invalid password for that case on the server seems fine too (and might be good to do regardless of this stuff), but I'd still like to fix the field-specific error message stuff in a cleaner way

Reply:
I haven't tried it yet, but I feel like having the submission function return _anything_ to the form might feel a bit awkward, especially since all it does now is just dispatches an action. But I also feel like our current representation of a form is only applicable until it gets submitted; after that, it doesn't seem like it should be form's responsibility anymore. Though I might be wrong about that.

But yeah, in any case I'm a bit stumped here and I don't see a clean way in which we can have validations _after_ the form gets submitted.

Reply:
I think if you can't figure out a clean-ish way to do it, then it's better to use the less optimal UX pattern (an error that isn't directly on the field) and leave a TODO to find that clean way. I don't think the UX difference is so great that it's going to be worth the problems it will cause if/when we make changes to the form stuff.